### PR TITLE
Making BulkRequest Releasable and updating server tests

### PR DESCRIPTION
**Pull request by masseyke**
This is the first step toward allowing bulk requests to reuse an underlying byte array rather than copying off smaller byte arrays for each request that later have to be garbage collected. In preparation for making BulkRequest implement RefCounted, I am updating it and BulkRequestBuilder to implement Releasable. Then I am updating all server tests to properly close those objects. In a subsequent PRs I will update other tests in the same way, and have BulkRequest implement RefCounted, updating the non-test code to properly use it.
Relates #104469

### CODE REVIEW

**Comment by masseyke** [comment]
@elasticmachine run elasticsearch-ci/part-3

File: server/src/main/java/org/elasticsearch/action/bulk/BulkRequest.java [hunk 1]
```diff
@@ -457,4 +459,7 @@ public long ramBytesUsed() {
     public Set<String> getIndices() {
         return Collections.unmodifiableSet(indices);
     }
+
+    @Override
+    public void close() {}
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/action/bulk/BulkRequest.java:464]: `BulkRequest` by virtue of being a transport request already has the ref-counted interface. We should use that and not add `Releasable`, the transport layer uses the ref counted interface on this thing for releasing requests eventually.

**Review by original-brownbear — CHANGES_REQUESTED** [review]
Sorry misclicked on the comment but see my comment inline, we can't use `Releasable` here.

File: server/src/main/java/org/elasticsearch/action/bulk/BulkRequest.java (hunk 1 above)
Comment (masseyke) [reply server/src/main/java/org/elasticsearch/action/bulk/BulkRequest.java:464]: Not having Releasable will make a lot of unit tests a lot more verbose though -- is that worth it?

Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/action/bulk/BulkRequest.java:464]: Yes it's the only way to get this right, we can't have `Releasable` mixed with ref-counting. At least when sending the response via a client that should result in it getting decremented by one, so for those use cases (which I expect to be the majority by a large margin) we shouldn't need any test code adjustment.

Comment (masseyke) [reply server/src/main/java/org/elasticsearch/action/bulk/BulkRequest.java]: We had an offline conversation about this. The summary is:

1. We will do explicit ref counting. So sending the response via a client will not result in in getting decremented.
2. We'll avoid using Releasable and try-with-resources as a convenience for calling decRef() in tests
3. We'll add a lot more utility methods and possibly a new RequestBuilder framework (like in #104497) to make the second point not quite as painful.

**Pull request closed without merging**